qutebrowser/browser/webengine/webenginetab.py

```diff
@@ -1237,6 +1243,133 @@ def _inject_site_specific_quirks(self):
+        log.webview.debug("Asking for Webauth user verification for "
+                          f"{request.relyingPartyId()}")
+        self._request = request
+        request.stateChanged.connect(self._on_ux_state_changed)
+        self._on_ux_state_changed(request.state())
+
+    def _on_ux_state_changed(self, state):
+        assert self._request is not None
+        log.webview.debug("Webauth UX state for "
+                          f"{self._request.relyingPartyId()}: {state}")
+        {
+            QWebEngineWebAuthUxRequest.WebAuthUxState.CollectPin:
+                self._ux_pin_request,
+            QWebEngineWebAuthUxRequest.WebAuthUxState.FinishTokenCollection:
+                lambda: message.info("Please touch your device now."),
+            QWebEngineWebAuthUxRequest.WebAuthUxState.SelectAccount:
+                self._ux_account_selection,
+            QWebEngineWebAuthUxRequest.WebAuthUxState.Completed:
+                self._ux_request_completed,
+            QWebEngineWebAuthUxRequest.WebAuthUxState.Cancelled:
+                self._ux_request_cancelled,
+            QWebEngineWebAuthUxRequest.WebAuthUxState.RequestFailed:
+                self._ux_request_failed
+        }[state]()
+
+    def _ux_pin_request(self):
+        assert self._request is not None
+        log.webview.debug("Collect Webuth pin for "
+                          f"{self._request.relyingPartyId()}")
+        answer = self._verification_required(self._request.relyingPartyId())
+        if answer is not None:
+            log.webview.debug("User verification accepted by user")
+            self._request.setPin(answer)
+        else:
+            log.webview.debug("User verification aborted by user")
+            self._request.cancel()
+
+    def _ux_account_selection(self):
+        assert self._request is not None
+        log.webview.debug("Select Webuth account for "
+                          f"{self._request.relyingPartyId()}")
+        answer = self._select_account(self._request.relyingPartyId(),
+                                      self._request.userNames())
+        if answer is not None:
+            log.webview.debug(f"Username selection accepted by user: {answer}")
+            self._request.setSelectedAccount(answer)
+        else:
+            log.webview.debug("Username selection aborted by user")
+            self._request.cancel()
+
+    def _ux_request_completed(self):
+        message.info("User verification completed.")
+        self._request = None
+
+    def _ux_request_cancelled(self):
+        message.info("User verification cancelled.")
+        self._tab.abort_questions.emit()
+        self._request = None
+
+    def _ux_request_failed(self):
+        assert self._request is not None
+        log.webview.debug("Webauth request failed for "
+                          f"{self._request.relyingPartyId()}: "
+                          f"{self._request.requestFailureReason()}")
+
+        reason = {
```

Comment:
Similar to above, we should ensure that we're actually covering all possible errors (even such added in the future) here. There are two ways to go about this:

- Also rewrite this as an `if` so that we can use `assert_never()` in the future (but I suppose for mapping the enum to strings, a dict makes a lot of sense)
- Move the dictionary into a class constant, and then add a test that makes sure the keys are equal to the enum members from Qt. If you search for `enum_members` in `tests/` you can get some inspiration there.